[supply] Uploading Multiple Apks

jzallas opened the pull request:
This is an extension to supply to allow multiple apk files to be uploaded to google play on the same release track. It is incredibly helpful for release gradle builds that use [apk splits](http://tools.android.com/tech-docs/new-build-system/user-guide/apk-splits) to generate multiple apks.

Here's an example of how to use it:
`supply -p com.example.app -u [apk_path_1],[apk_path_2] -a alpha`

Disclaimer: I'm not _that_ familiar with ruby so if anything needs to be adjusted, feel free to advise.

jzallas commented:
Looking through other PRs, I saw this:
> Also, it looks like your PR got started before we fixed some trouble with our CI setup. In order to fix your build please run:
> 
> `git cherry-pick f6a7d2100b639c27ecd70c51f59f3542b9dc5cea`
> before squashing to bring the fix into your branch. Thanks :+1:

I'm guessing I might need to do this also?

asfalcone commented on supply/lib/supply/options.rb:
```diff
@@ -68,12 +68,19 @@ def self.available_options
                                        UI.user_error! "Could not find apk file at path '#{value}'" unless File.exist?(value)
                                        UI.user_error! "apk file is not an apk" unless value.end_with?(value)
                                      end),
+        FastlaneCore::ConfigItem.new(key: :apk_dir,
```
Any good `default_value` you can think of for this ConfigItem?

jzallas replied:
`default_value: Dir[File.join("app", "build", "outputs", "apk")].last`
This is the only thing I can think of but it could produce unwanted results if unsigned builds are left in that directory. Is a default required?

asfalcone replied:
defaults are not required but it is always nice to provide if there is something that will work for most people.

asfalcone replied:
Another option is to make this param actually be an array of apk paths instead of a directory and you could use the `GRADLE_ALL_APK_OUTPUT_PATHS` output from the [gradle](https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/gradle.rb) action as the default!

jzallas replied:
I'll try to get to this before the end of the week~

asfalcone commented on supply/lib/supply/uploader.rb:
```diff
@@ -83,20 +87,45 @@ def upload_screenshots(language)
+    def upload_binaries
+      apk_dir = Supply.config[:apk_dir]
+
+      UI.user_error!("Could not find folder #{apk_dir}") unless File.directory? apk_dir
```
Move this checking logic to `verify_block` in ConfigItem

asfalcone commented on supply/lib/supply/uploader.rb:
```diff
@@ -83,20 +87,45 @@ def upload_screenshots(language)
+
+      UI.user_error!("Could not find folder #{apk_dir}") unless File.directory? apk_dir
+
+      apk_files = Dir.foreach(apk_dir).sort { |x, y| x <=> y }
```
What is the significance of sorting the apk files?

asfalcone commented:
This PR looks really great @jzallas! I've added a few comments that need to be addressed and I am looking into the CI failure. Rebasing on master might help. :+1:

jzallas commented:
- Moved verification logic into ConfigItem
- Removed unnecessary sorting
- Changed input from directory to array of paths
- Added default array from `GRADLE_ALL_APK_OUTPUT_PATHS`

Rebased off of master and squashed everything into one commit!

asfalcone commented on supply/lib/supply/options.rb:
```diff
@@ -68,12 +68,27 @@ def self.available_options
+                                       UI.user_error!("Could not evaluate array from '#{value}'") unless value.kind_of?(Array)
+                                       value.each do |path|
+                                         UI.user_error! "Could not find apk file at path '#{path}'" unless File.exist?(path)
+                                         UI.user_error! "apk file is not an apk" unless path.end_with?(path)
```
I think `path.end_with?(path)` is probably a typo here, as well as `value.end_with?(value)` that it was mirrored from. These should be doing `.end_with?('.apk')` right?

jzallas replied:
yep, that makes so much more sense :+1:

asfalcone commented on supply/lib/supply/uploader.rb:
```diff
@@ -83,20 +87,39 @@ def upload_screenshots(language)
       end
     end
 
+    def upload_binaries
```
could this code be tighted up a little bit? maybe one `upload_binaries` method that expects an array and in the single apk case we just pass it `[Supply.config[:apk_path]]`?

asfalcone commented:
@jzallas This is really looking good and this build is passing now! We're getting very close to merge-town. I left 2 more comments for ya :+1:

jzallas commented:
@asfalcone added your requested changes~

1. corrected `end_with?(...)` to use `'.apk'` on both single and multi apk options
2. `upload_binaries` now handles both single and multi apk scenarios
3. removed `upload_binary` used for single apk since it is no longer needed

Commit:
Deploying Multiple APKs

	Added upload_binaries method which will upload multiple binaries and assign all the uploaded binaries to a single track	perform_upload now checks if apk_paths were provided and follows the multi apk logic
	Refactored the binary upload logic to be reused for multiple apk uploads
	Extracted logic that handles track uploads (to be reused for multiple apk)
	client's update_track now handles multiple version codes
	Added an option to use an array of apk file paths
	Hooked GRADLE_ALL_APK_OUTPUT_PATHS as a default value for apk_paths param
	Removed upload_binary
	Refactored upload_binaries to handle both single and multi apk scenarios

asfalcone commented on supply/lib/supply/uploader.rb:
```diff
@@ -105,13 +117,23 @@ def upload_binary
         Helper.log.info "No apk file found, you can pass the path to your apk using the `apk` option"
       end
+
+      UI.user_error!("Could not upload #{apk_path}") unless apk_version_code
```
I think this line needs to move up inside the `if apk_path` branch. My test setup didn't have an apk path param (but also didn't have skip_upload_apk) and the behavior from the previous version of the tool changes such that I get the "Could not upload" error. Your code here might be more correct but let's not introduce a behavior change

jzallas replied:
Moved the error up into the if branch. I think it's actually better here because it's raised the second we know it failed instead of trying to upload obbs and changelogs with the wrong `apk_version_code`
```
Helper.log.info "Preparing apk at path '#{apk_path}' for upload..."
apk_version_code = client.upload_apk(apk_path)
UI.user_error!("Could not upload #{apk_path}") unless apk_version_code
```

asfalcone commented:
@jzallas Just did a final pass with testing and discovered one issue that needs to be fixed. Please checkout that comment. This branch also has conflicts with master. Then I think we're ready to go! :rocket:

jzallas commented:
Moved that error code and resolved conflicts :+1:

asfalcone commented:
Awesome work @jzallas :tada:

lacostej commented:
@jzallas out of curiosity, do you know if this solves #1962 ?

jzallas commented:
@lacostej yes, it was built specifically for that purpose. :+1:
